lib/event_dispatcher/dispatcher.rb

```diff
@@ -0,0 +1,104 @@
+module EventDispatcher
+  class Dispatcher
+    private
+
+    attr_reader :event_listeners
+    attr_reader :sorted
+
+    public
+
+    def initialize
+      @event_listeners = {}
+      @sorted = {}
+    end
+
+    def dispatch(event_name:, event: EventDispatcher::BaseEvent.new)
+      listeners = listeners(event_name: event_name)
+
+      do_dispatch(listeners: listeners, event: event) if listeners
+
+      event
+    end
+
+    def listeners(event_name: nil)
+      unless event_name.nil?
+        return [] if !event_listeners.key?(event_name) || event_listeners[event_name].empty?
+
+        sort_listeners(event_name: event_name) if sorted[event_name].nil? || sorted[event_name].empty?
+
+        return sorted[event_name]
+      end
+
+      event_listeners.each_key do |key|
+        sort_listeners(event_name: key) if sorted[key].nil? || sorted[key].empty?
+      end
+
+      sorted
+    end
+
+    def listeners?(event_name: nil)
+      unless event_name.nil?
+        return event_listeners.key?(event_name) && !event_listeners[event_name].empty?
+      end
+
+      event_listeners.each { |listener| return true if listener }
```

Comment:
With some adjustments (you delete any keys whenever there are no more listeners for them), this could just be `event_listeners.empty?`).